### PR TITLE
Add abstraction WIP

### PR DESCRIPTION
**Pull request by BlueBlueBlob**
## Description

It's draft. I need comments to continue.

#80 

## Checklist
  - [x] The code change is tested and works locally.
  - [x] tests pass. **Your PR won't be merged unless tests pass**
  - [x] There is no commented out code in this PR

### CODE REVIEW

**Comment by BlueBlueBlob** [comment]
It's stock.
I'm not sure, do you want product related functions in Product class ? For me it seems to be a good idea : oriented object.

**Comment by BlueBlueBlob** [comment]
And I did minimal test so coverage and others checks will fail. It's a draft.

**Comment by BlueBlueBlob** [comment]
@markrussel

File: pygrocy/grocy.py [hunk 1]
```diff
@@ -159,64 +156,65 @@ def last_done_by(self) -> UserDto:
...
-    def product(self, product_id: int) -> ProductDetailsResponse:
-        return self._api_client.get_product(product_id)
+class Stock():
+    def __init__(self, api_client: GrocyApiClient):
+        self._api_client = api_client
+        self._products_list = self._get_stock()
+        self._volatile_products = self._get_volatile_stock()
+        self._expiring_products = self._get_expiring_products()
+        self._expired_products = self._get_expired_products()
+        self._missing_products = self._get_missing_products()
```
[the remark this answers is not in the document]
Comment (BlueBlueBlob) [reply pygrocy/grocy.py]: It's only one request. Before it was 3 same request. So I don't understand your point.

File: pygrocy/grocy.py [hunk 2]
```diff
@@ -159,64 +156,65 @@ def last_done_by(self) -> UserDto:
...
-        expired_product = [Product(resp) for resp in raw_expired_product]
+    def add_missing_product_to_shopping_list(self):
+        return self._api_client.add_missing_product_to_shopping_list()
 
-        if get_details:
-            for item in expired_product:
-                item.get_details(self._api_client)
-        return expired_product
+    def add_product_to_shopping_list(self, product_id: int,  amount: int = 1):
+        return self._api_client.add_product_to_shopping_list(product_id, amount)
```
[the remark this answers is not in the document]
Comment (BlueBlueBlob) [reply pygrocy/grocy.py]: Yeah it's just cut and paste for now.

File: pygrocy/grocy.py [hunk 3]
```diff
@@ -159,64 +156,65 @@ def last_done_by(self) -> UserDto:
...
+            for item in shopping_list:
                 item.get_details(self._api_client)
-        return expiring_product
+        return shopping_list
 
-    def expired_products(self, get_details: bool = False) -> List[Product]:
-        raw_expired_product = self.volatile_stock().expired_products
-        expired_product = [Product(resp) for resp in raw_expired_product]
+    def add_missing_product_to_shopping_list(self):
+        return self._api_client.add_missing_product_to_shopping_list()
```
[the remark this answers is not in the document]
Comment (BlueBlueBlob) [reply pygrocy/grocy.py]: Not sure. Removing return will raise an error in cofefactor. I will check

File: pygrocy/grocy.py [hunk 4]
```diff
@@ -159,64 +156,65 @@ def last_done_by(self) -> UserDto:
...
-        stock = [Product(resp) for resp in raw_stock]
-        return stock
-
-    def volatile_stock(self) -> CurrentVolatilStockResponse:
-        return self._api_client.get_volatile_stock()
-
-    def expiring_products(self, get_details: bool = False) -> List[Product]:
-        raw_expiring_product = self.volatile_stock().expiring_products
-        expiring_product = [Product(resp) for resp in raw_expiring_product]
+    def _get_shopping_list(self, get_details: bool = False) -> List[ShoppingListProduct]:
```
[the remark this answers is not in the document]
Comment (BlueBlueBlob) [reply pygrocy/grocy.py]: Yes for the refresh. For all "list" class I will add it.

File: pygrocy/grocy.py [hunk 5]
```diff
@@ -124,16 +117,20 @@ def product(self) -> Product:
...
         self._id = raw_chore.chore_id
         self._last_tracked_time = raw_chore.last_tracked_time
         self._next_estimated_execution_time = raw_chore.next_estimated_execution_time
 
         self._name = None
         self._last_done_by = None
 
-    def get_details(self, api_client: GrocyApiClient):
-        details = api_client.get_chore(self.id)
+    def execute_chore(self, done_by: int = None, tracked_time: datetime = datetime.now()):
```
[the remark this answers is not in the document]
Comment (BlueBlueBlob) [reply pygrocy/grocy.py]: Cut and paste 😹

File: pygrocy/grocy.py [hunk 6]
```diff
@@ -124,16 +117,20 @@ def product(self) -> Product:
 
 
 class Chore(object):
-    def __init__(self, raw_chore: CurrentChoreResponse):
+    def __init__(self, api_client: GrocyApiClient, raw_chore: CurrentChoreResponse):
```
[the remark this answers is not in the document]
Comment (BlueBlueBlob) [reply pygrocy/grocy.py]: Diff between chores and shopping list :
Chores have only functions related to chore.
Shopping list have functions related to product in the list AND functions related to the list itself.

https://demo.grocy.info/api#/Chores

File: pygrocy/grocy.py [hunk 7]
```diff
@@ -159,64 +156,65 @@ def last_done_by(self) -> UserDto:
...
+        self._expired_products = self._get_expired_products()
+        self._missing_products = self._get_missing_products()
 
-    def chores(self, get_details: bool = False) -> List[Chore]:
-        raw_chores = self._api_client.get_chores()
-        chores = [Chore(chore) for chore in raw_chores]
+    def _get_stock(self) -> List[Product]:
+        raw_stock = self._api_client.get_stock()
+        stock = [Product(resp) for resp in raw_stock]
+        return stock
```
[the remark this answers is not in the document]
Comment (BlueBlueBlob) [reply pygrocy/grocy.py]: The second solution can be done trough the Grocy SQL database which is not exposed through API. So... Forget it.

For the first one : tuples and refresh method will do job. I already use on another project, works nice.

**Comment by BlueBlueBlob** [comment]
So why sometimes get_details are used or not :
Shopping list : a product in the list can have a product_id or not 😹 and the way to retrieve the list is different...

Chore : because we need one field only :/

Product : not anymore since stock returns it.
But a missing product is different from a Product...

Grocy API itself is not consistent. I do my best 😹

**Comment by markusressel** [comment]
> So why sometimes get_details are used or not :
> Shopping list : a product in the list can have a product_id or not joy_cat and the way to retrieve the list is different...
> 
> Chore : because we need one field only :/
> 
> Product : not anymore since stock returns it.
> But a missing product is different from a Product...
> 
> Grocy API itself is not consistent. I do my best joy_cat

Yeah no worries, just adding my thoughts as I go along. If something seems offensive, its definitely not meant that way :smile:

I do agree, the Grocy API is quite a mess. Sadly, from what I read so far in the Grocy issues, the maintainer doesn't seem to be that open to big changes, so I am not sure if its possible to change the API, plus it would probably also need a lot of work on the frontend side. I guess it is what it is.

**Comment by BlueBlueBlob** [comment]
Hey don't worry, me AND my English are rude. Sorry.

But seriously I don't like how we are working.
For me it's not team working but just my worst customer.  But I will try to more explain my work.

You have a good ideas sorry for my answers

File: pygrocy/grocy.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BlueBlueBlob) [reply pygrocy/grocy.py]: Sorry I checked on my code where it was already corrected.

**Comment by BlueBlueBlob** [comment]
Sorry but I'm done.

I will focus on another project.

Good luck.

Best regards.

**Comment by SebRut** [comment]
> But seriously I don't like how we are working.
For me it's not team working but just my worst customer. 

> Sorry but I'm done.
I will focus on another project.


Really sad to hear you leaving this project. I always appreciated your work and motivation. Good luck :)

**Pull request closed without merging**